### PR TITLE
Replacing hal.ex.shared_device with hal.devices.* ops.

### PR DESCRIPTION
**Pull request by benvanik**
This bumps the HAL version as the existing hal.ex.shared_device op is removed. The runtime HAL module still only reports a single device and the compiler code is still selecting the "default" device (0). Future changes will support multiple `--device=` flags to register multiple available devices in tools and add attributes/lookup for mapping stream/device affinity to device ordinals.

The plan is to have a `hal.devices.lookup` op that is hoisted into an initializer to enumerate devices and store the selected ones in globals. The enumeration code will emit a loop over `hal.devices.count` and use `hal.devices.get` and the various device query methods to match the requested devices.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 1]
```diff
@@ -1863,6 +1845,57 @@ def HAL_DeviceQueueFlushOp : HAL_Op<"device.queue.flush"> {
...
+}
+
+def HAL_DevicesGetOp : HAL_PureOp<"devices.get", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+]> {
+  let summary = [{returns the device with the given index}];
+  let description = [{
+    Returns the device with the given index in the [0, hal.devices.count) range.
+    Devices may be lazily initialized upon first use.
+  }];
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td:1883]: > The plan is to have a `hal.devices.lookup` op that is hoisted into an initializer to enumerate devices and store the selected ones in globals. The enumeration code will emit a loop over `hal.devices.count` and use `hal.devices.get` and the various device query methods to match the requested devices.

So will the HAL dialect provide ways to see details about each device, such as supported executable formats and this info? https://github.com/openxla/iree/blob/3654fd683c1cbb5df76023dd0880ca164fba55c8/runtime/src/iree/hal/device.h#L72-L80

`hal.device.query` (reference: https://iree.dev/reference/mlir-dialects/HAL/#haldevicequery-haldevicequeryop) doesn't look like enough on its own.

If the compiler doesn't actually need that information, what will it do instead? I imagine what the compiler needs ahead of time is pretty different from what a programmer would need. If I were writing code against this, I would want to loop over all devices and find "the discrete GPU device", for example.

File: runtime/src/iree/modules/hal/module.c [hunk 2]
```diff
@@ -32,97 +32,143 @@
 // Module type definitions
 //===----------------------------------------------------------------------===//
 
-#define IREE_HAL_MODULE_VERSION_0_1 0x00000001u
-#define IREE_HAL_MODULE_VERSION_LATEST IREE_HAL_MODULE_VERSION_0_1
+#define IREE_HAL_MODULE_VERSION_0_2 0x00000002u
+#define IREE_HAL_MODULE_VERSION_LATEST IREE_HAL_MODULE_VERSION_0_2
```
Comment (ScottTodd) [inline runtime/src/iree/modules/hal/module.c:36]: > This bumps the HAL version as the existing hal.ex.shared_device op is removed. The runtime HAL module still only reports a single device and the compiler code is still selecting the "default" device (0). Future changes will support multiple `--device=` flags to register multiple available devices in tools and add attributes/lookup for mapping stream/device affinity to device ordinals.

Do you expect this version to change a few more times as part of this work? If so, would it make sense to either develop the changes on a branch or to plan releases around these changes (e.g. push a stable release before this lands, then another after a few patches land).

Comment (benvanik) [reply runtime/src/iree/modules/hal/module.c:36]: I don't believe so - I think with count/get and the query call we can do everything we need - query takes strings and we can easily add new queries without changing the HAL API itself.

File: compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td:1883]: device query already exposes the ability to check for executable format support and I'll be adding additional queries as needed

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td:1883]: Oh, great :D.

I'd forgotten that `hal.executable.format` was a "well-known key" (should have read the page I linked...). I was thinking that the keys would be device-specific and common queries would have dedicated methods.

**Review by ScottTodd — APPROVED** [review]

**Comment by qedawkins** [comment]
> The plan is to have a hal.devices.lookup op that is hoisted into an initializer to enumerate devices and store the selected ones in globals. The enumeration code will emit a loop over hal.devices.count and use hal.devices.get and the various device query methods to match the requested devices.

Would you be willing to write up an issue and/or hold a mini design meeting for where this is expected to end up? I'd like to understand the bounds of what this is scoped for. (Or I can just ask my questions directly somewhere).

**Comment by benvanik** [comment]
The TLDR is that most of it will overlap with what we are doing for executable condition regions: emitting queries based on attributes that are memoized/hoisted. The existing `#hal.affinity.queue` attribute will be replaced with one that allows for specifying a `#hal.device.target` attribute, an ordinal, or some existing device in addition to a queue. `hal.device.targets` will likely need to change to support device sets ("run on device A _and_ B _or_ device C" etc). Will probably start small as we can iterate on this stuff much quicker once it's just compiler changes - hopefully this PR (and some additional queries added on demand) are the only things that cross the runtime boundary.

Would love to chat/brainstorm, but there's no design yet just gut feelings: this is one of those tricky things where by the time I can articulate what it looks like in an issue/design it'll have been implemented and just be a PR description :)

**Comment by antiagainst** [comment]
Cool stuff! RIP `hal.ex.shared_device`! Thanks for the long services! :)

**Comment by qedawkins** [comment]
> The existing #hal.affinity.queue attribute will be replaced with one that allows for specifying a #hal.device.target attribute, an ordinal, or some existing device in addition to a queue. hal.device.targets will likely need to change to support device sets ("run on device A and B or device C" etc).

Unsurprisingly you are pre-empting my questions :D

> Would love to chat/brainstorm, but there's no design yet just gut feelings: this is one of those tricky things where by the time I can articulate what it looks like in an issue/design it'll have been implemented and just be a PR description :)

Maybe tomorrow or next week chatting would be awesome if you're open to it. Don't want to prefetch/stress too hard on partial designs though.
